// @atlaskit/button/dist/cjs/button-group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ButtonGroup;
exports.groupItemStyles = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@emotion/core");

var _constants = require("@atlaskit/theme/constants");

/** @jsx jsx */
var groupItemStyles = {
  flex: '1 0 auto',
  display: 'flex',

  /* margins don't flip when the layout uses dir="rtl", whereas pseudos do */
  '& + &::before': {
    content: "''",
    display: 'inline-block',
    width: "".concat((0, _constants.gridSize)() / 2, "px")
  }
};
exports.groupItemStyles = groupItemStyles;

function ButtonGroup(_ref) {
  var appearance = _ref.appearance,
      children = _ref.children;
  return (0, _core.jsx)("div", {
    css: {
      display: 'inline-flex'
    }
  }, _react.default.Children.map(children, function (child, idx) {
    if (!child) {
      return null;
    }

    return (0, _core.jsx)("div", {
      key: idx,
      css: groupItemStyles
    }, appearance ? /*#__PURE__*/_react.default.cloneElement(child, {
      appearance: appearance
    }) : child);
  }));
}
